File: routes/queue.js
const express = require('express');
const { acquireLock, releaseLock, getQueue } = require('./queue-manager');
const wsManager = require('../websocket');
const router = express.Router();

// Add user to queue
router.post('/add', async (req, res) => {
  try {
    const { piId, printerId, userId } = req.body;

    if (!piId || !printerId || !userId) {
      return res.status(400).json({ error: 'Missing required parameters' });
    }

    const success = await acquireLock(req.db, piId, printerId, userId);
    
    // Broadcast real-time update
    const updatedQueue = await getQueue(req.db, piId, printerId);
    wsManager.broadcastQueueUpdate(piId, printerId, updatedQueue);
    
    res.json({ success });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Release user from queue
router.post('/release', async (req, res) => {
  try {
    const { piId, printerId, userId } = req.body;

    if (!piId || !printerId || !userId) {
      return res.status(400).json({ error: 'Missing required parameters' });
    }

    const success = await releaseLock(req.db, piId, printerId, userId);
    
    // Broadcast real-time update
    const updatedQueue = await getQueue(req.db, piId, printerId);
    wsManager.broadcastQueueUpdate(piId, printerId, updatedQueue);
    
    res.json({ success });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get queue status
router.get('/status', async (req, res) => {
  try {
    const { piId, printerId } = req.query;

    if (!piId || !printerId) {
      return res.status(400).json({ error: 'Missing piId or printerId' });
    }

    console.log('Getting queue for:', { piId, printerId });
    
    // Check if database is connected
    if (!req.db) {
      console.error('Database not connected');
      return res.status(500).json({ error: 'Database not connected' });
    }
    
    const queue = await getQueue(req.db, piId, printerId);
    console.log('Queue result:', queue);
    
    const safeQueue = Array.isArray(queue) ? queue : [];
    console.log('Sending response:', { queue: safeQueue });
    res.json({ queue: safeQueue });
  } catch (error) {
    console.error('Queue status error:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: routers/queue.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from pymongo.database import Database
from queue_manager import acquire_lock, release_lock, get_queue
from websocket_manager import ws_manager

router = APIRouter()

# Pydantic models
class QueueAdd(BaseModel):
    piId: str
    printerId: str
    userId: str

class QueueRelease(BaseModel):
    piId: str
    printerId: str
    userId: str

# Dependency to get database
def get_database():
    from main import db
    if db is None:
        raise HTTPException(status_code=500, detail="Database not connected")
    return db

@router.post("/add")
async def add_to_queue(queue_data: QueueAdd, db: Database = Depends(get_database)):
    try:
        success = acquire_lock(db, queue_data.piId, queue_data.printerId, queue_data.userId)
        
        # Broadcast real-time update
        updated_queue = get_queue(db, queue_data.piId, queue_data.printerId)
        await ws_manager.broadcast_queue_update(queue_data.piId, queue_data.printerId, updated_queue)
        
        return {"success": success}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/release")
async def release_from_queue(queue_data: QueueRelease, db: Database = Depends(get_database)):
    try:
        success = release_lock(db, queue_data.piId, queue_data.printerId, queue_data.userId)
        
        # Broadcast real-time update
        updated_queue = get_queue(db, queue_data.piId, queue_data.printerId)
        await ws_manager.broadcast_queue_update(queue_data.piId, queue_data.printerId, updated_queue)
        
        return {"success": success}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/status")
async def get_queue_status(piId: str, printerId: str, db: Database = Depends(get_database)):
    try:
        print(f"Getting queue for: {piId}, {printerId}")
        
        if not db:
            print("Database not connected")
            raise HTTPException(status_code=500, detail="Database not connected")
        
        queue = get_queue(db, piId, printerId)
        print(f"Queue result: {queue}")
        
        safe_queue = queue if isinstance(queue, list) else []
        print(f"Sending response: {safe_queue}")
        return {"queue": safe_queue}
    except Exception as e:
        print(f"Queue status error: {e}")
        raise HTTPException(status_code=500, detail=str(e))